### DISCUSSION TITLE
Saved Image is not equal to the original

### DISCUSSION THREAD

**Discussion by cgrain** [Project-MONAI/MONAI#2029] [Q&A]
**Intro**
The other day, I tried to speed up the training process of our project by saving training images after the deterministic transformations had been executed. (Resampling in particular). I encountered the following problem:

**Expectation**
Saved images are equal or almost equal (`np.allclose(X, X_hat) == True`)

**Procedure**
1. Loading images
2. Performing transformations (It turned out, the culprit is the resampling procedure) and getting **X**
3. save image **X** to _file_ but also keep it in memory. By default, it resamples the image back to the original affine_matrix. So, turn resampling off. It is possible to delete the original affine matrix, but that is not preferable. 
4. load _file_ and store it in **X_hat**
5. Check for equality, turns out, it is not. 

**Concluding thoughts**
It seems that the orientation changes during saving, even if resampling is off. Is this by design, or is this a bug? I have tried all kinds of saving images, i.e. `NiftiSaver`, the transformation, etcetera. idem for the loader, i.e. `LoadImage`, `ImageDataset`, etcetera. Probably, they all share the same implementation, so that does not really help. 
kind regards,

**Code**
```python
import numpy as np
import monai


def create_random_img():
    Random = np.random.default_rng()
    image = Random.standard_normal((256, 256, 35))
    original_affine = np.asarray(
        [
            [1.1e-2, -4e-3, 4.3, 3.3e1],
            [-7e-1, -1e-2, 6.5e-2, 2.1e2],
            [-1e-2, 7e1, 2.9e-2, -1.9e2],
            [0, 0, 0, 1],
        ]
    )  # Copied from a dataset (So a valid sample!)
    original_file_name = "Test.nii.gz"
    meta_data = {
        "affine": original_affine,
        "original_affine": original_affine,
        "filename_or_obj": original_file_name,
    }
    return {"img": image, "img_meta_dict": meta_data}


def resample_image(image):
    RESAMPLERD = monai.transforms.SpacingD(
        keys=["img"], pixdim=[1, 1, 12], diagonal=True
    )
    return RESAMPLERD(image)


def save_image(image):
    SAVER = monai.data.NiftiSaver(resample=False)
    SAVER.save(image["img"], meta_data=image["img_meta_dict"])
    new_file_name = "./Test/Test_seg.nii.gz"  # Trust me on this
    return image, new_file_name


def load_image(file_name):
    LOADER = monai.transforms.LoadImage()
    return LOADER(file_name)


def main():
    image = create_random_img()
    resampled_img = resample_image(image)
    X = resampled_img
    X["img"] = np.expand_dims(X["img"], 0)

    _, file_name = save_image(X)
    X_hat = load_image(file_name)

    # This should (imo) not be necessary, but let's give it a go
    image_hat = np.moveaxis(X_hat[0], [0, 2], [2, 0])

    print(np.allclose(X["img"], image_hat))  # Expecting True (It is False)


if __name__ == "__main__":
    main()
```

**Comment by wyli** [comment]
I'm converting this to a discussion as it is mainly a usage issue:
- SpacingD assumes "channel-first" format, so `image` should be `Random.standard_normal((1, 256, 256, 35))`
- the `niftisaver` will assume the current data array is in the `['img_meta_dict']['affine']` coordinates, and try to convert the array to the `['img_meta_dict']['orignal_affine']` coordinates, the only way to prevent this change to the array is to set `original_affine` = `affine`

put them together, here is a working example:
```py
import numpy as np
import monai


def create_random_img():
    Random = np.random.default_rng()
    image = Random.standard_normal((1, 256, 240, 35))
    original_affine = np.asarray(
        [
            [1.1e-2, -4e-3, 4.3, 3.3e1],
            [-7e-1, -1e-2, 6.5e-2, 2.1e2],
            [-1e-2, 7e1, 2.9e-2, -1.9e2],
            [0, 0, 0, 1],
        ]
    )  # Copied from a dataset (So a valid sample!)
    original_file_name = "Test.nii.gz"
    meta_data = {
        "affine": original_affine,
        "original_affine": original_affine,
        "filename_or_obj": original_file_name,
        "spatial_shape": image.shape[1:],
    }
    return {"img": image, "img_meta_dict": meta_data}


def resample_image(image):
    RESAMPLERD = monai.transforms.SpacingD(
        keys=["img"], pixdim=[1, 1, 12], diagonal=True
    )
    return RESAMPLERD(image)


def save_image(image):
    SAVER = monai.data.NiftiSaver(resample=True)
    image['img_meta_dict']['original_affine'] = image['img_meta_dict']['affine']
    SAVER.save(image["img"], meta_data=image["img_meta_dict"])
    new_file_name = "./Test/Test_seg.nii.gz"  # Trust me on this
    return image, new_file_name


def load_image(file_name):
    LOADER = monai.transforms.LoadImage()
    return LOADER(file_name)


def main():
    image = create_random_img()
    resampled_img = resample_image(image)
    X = resampled_img

    _, file_name = save_image(X)
    X_hat = load_image(file_name)
    image_hat = X_hat[0]

    print(np.allclose(X["img"], image_hat))  # Expecting True


if __name__ == "__main__":
    main()
```

**Reply by cgrain** [reply]
Thank you for your reply. Your second tip indeed works (and I have already implemented something similar) but it feels hacky. My assumption of saving anything is (not necesarily a nifti file): 
```
X = any data
Save(X, file_name)
X_hat = Load(file_name)
assert X == X_hat
```
In this case, that does not happen. If you do not change the data in the exact format first, you do not get exactly the data you tried to save back. My main reason for creating an issue was that I am wondering _why_ the NiftiSaver changes your data even if you set resampling to False. It seems to me, that by setting resampling to False, you want to save the transformed data, whether that data is because you resampled, changed the orientation or anything else. What am I missing? 
P.S. 
at the moment, I cannot include a `SaveImage` transform in a simple `Compose` and expect the data to be the same. That is one of the reasons it feels hacky to me :)

**Reply by wyli** [reply]
sure, I feel this component is mainly for saving the data in (or close to) their original coordinate system. if you want to save an image as it is, perhaps nibabel's Nifti1Image.save method is good enough.

also, thanks for the suggestions, as we now have the invertible spacing and orientation transforms, we are planning to simplify the save/load components
